### ISSUE TITLE
Custom test: check import of library

### ISSUE THREAD

**Issue by martinkotlan** [koalaman/shellcheck#2460]
Please, is there some way to check the presence of import of certain library in a script? Non-presence should emit warning message. I have already written a recursive function, that checks the root's children nodes for the one that I want to check for, but it's not clear to me, how to add it to Custom.hs file, because DevGuide mentions just editing Analytics.hs file and my Haskell knowledge is fairly limited. I would be grateful for a advice, thank you.

**Comment by koalaman** [comment]
Sure, the only slight issue is that triggering on non-presence by default will break a whole bunch tests that now have unexpected warnings in them, so it may be useful to have it as an optional check instead. Here's an example where you can obviously replace `verifyImport` with a more targeted implementation:

```
-- This is Custom.hs
{-# LANGUAGE TemplateHaskell #-}
module ShellCheck.Checks.Custom (checker, ShellCheck.Checks.Custom.runTests) where

import Control.Monad.Writer
import Data.Maybe
import ShellCheck.AST
import ShellCheck.ASTLib
import ShellCheck.AnalyzerLib
import ShellCheck.Interface
import Test.QuickCheck

checker :: AnalysisSpec -> Parameters -> Checker
checker spec _ = checksFor spec

checksFor spec =
    if "require-mylib" `elem` asOptionalChecks spec
    then enabledChecker
    else disabledChecker

enabledChecker =
    Checker {
        perScript = verifyImport,
        perToken = const $ return ()
      }

disabledChecker =
    Checker {
        perScript = const $ return (),
        perToken = const $ return ()
      }

triggersOn = producesComments enabledChecker


prop_verifyImport1 = triggersOn "# shellcheck enable=require-mylib\necho 'Hello World'" == Just True
prop_verifyImport2 = triggersOn "# shellcheck enable=require-mylib\nsource mylib\necho 'Hello World'" == Just False

verifyImport :: Root -> Analysis
verifyImport (Root t) = do
    unless (any isRelevantImport flatListOfTokens) $
        err (getId t) 9001 "Did not find 'source mylib' in this script"
  where
    flatListOfTokens :: [Token]
    flatListOfTokens = execWriter $ doAnalysis (\t -> tell [t])  t

    isRelevantImport :: Token -> Bool
    isRelevantImport t =
        case t of
            T_SimpleCommand id assignments (cmd:arg:_) -> fromMaybe False $ do
                cmdString <- getLiteralString cmd
                argString <- getLiteralString arg
                return $ cmdString == "source" && argString == "mylib"
            _ -> False


return []
runTests = $quickCheckAll
```

To get access to the list of requested optional checks, you'll have to kludge Analyzer.hs to pass it along to Custom.hs the way it's already done for Commands.hs:

```
--- a/src/ShellCheck/Analyzer.hs
+++ b/src/ShellCheck/Analyzer.hs
@@ -42,7 +42,7 @@ analyzeScript spec = newAnalysisResult {
 
 checkers spec params = mconcat $ map ($ params) [
     ShellCheck.Checks.Commands.checker spec,
-    ShellCheck.Checks.Custom.checker,
+    ShellCheck.Checks.Custom.checker spec,
     ShellCheck.Checks.ShellSupport.checker
     ]
```
